### PR TITLE
Expose Consul template configuration parameters

### LINKED ISSUE #2623: Nomad job using template stanza permanently dead after transient consul errors
### Nomad version
Output from `nomad version`
```
Nomad v0.5.6
```
### Operating system and Environment details
Ubuntu 16.04 LTS

### Issue
Every week my job reproducably turns "dead":

```
ID            = scorecard
Name          = scorecard
Type          = service
Priority      = 50
Datacenters   = cbk1
Status        = dead
Periodic      = false
Parameterized = false

Summary
Task Group  Queued  Starting  Running  Failed  Complete  Lost
web         2       0         0        0       0         0

Allocations
No allocations placed
```

I am not sure yet what the cause is, that's why I provide so much info and logs here.

I suspect it could be connected with the fact that all nodes in the nomad cluster reboot regularly, automatically.

Nomad cluster looks like this:
```
syseleven@nomad-servicehost1:~$ consul members
Node                 Address            Status  Type    Build  Protocol  DC
nomad-loadbalancer0  192.168.2.13:8301  alive   client  0.8.1  2         cbk1
nomad-node0          192.168.2.15:8301  alive   client  0.8.1  2         cbk1
nomad-node1          192.168.2.14:8301  alive   client  0.8.1  2         cbk1
nomad-servicehost0   192.168.2.11:8301  alive   server  0.8.1  2         cbk1
nomad-servicehost1   192.168.2.10:8301  alive   server  0.8.1  2         cbk1
nomad-servicehost2   192.168.2.12:8301  alive   server  0.8.1  2         cbk1
```

servicehosts are consul and nomad servers, the rest are consul and nomad clients.

### Reproduction steps
Not sure yet

### Nomad Server & Client logs (if appropriate)
see attachment 
[scorecard.log.txt](https://github.com/hashicorp/nomad/files/980905/scorecard.log.txt)

### Job file (if appropriate)
```
job "scorecard" {
  region      = "global"
  datacenters = ["cbk1"]
  priority    = 50

  type = "service"

  group "web" {
    count = 2

    task "scorecard-nginx" {
      driver = "docker"

      config {
        image = "nginx:latest"
        volumes = [
          "default.conf:/etc/nginx/conf.d/default.conf",
          "public_html:/var/www"
        ]
        port_map {
          http = 80
        }
      }

      service {
        name = "scorecard"
        port = "http"
        check {
          name     = "alive"
          type     = "tcp"
          interval = "10s"
          timeout  = "2s"
        }
      }

      resources {
        cpu    = 2000 # 2 GHz
        memory = 256 # 1GB
        network {
          mbits = 10
          port "http" {
          }
        }
      }

      template {
        destination   = "default.conf"
        change_mode   = "restart"
        data          = <<EOH

                        server {
                            listen       80;
                            server_name  localhost;

                            error_page   500 502 503 504  /50x.html;
                            index index.html;
                            location / {
                                root   /var/www;
                            }
                            location = /50x.html {
                                root   /usr/share/nginx/html;
                            }
                        }
                        EOH
      }

      artifact {
        source = "{{ HTTPS_ARTIFACT_URL }}/scorecard/public_html.tgz"
        destination = "."
      }

      template {
        source        = "public_html/index.html.tpl"
        destination   = "public_html/index.html"
        change_mode   = "noop"
      }
    }
  }
}
```

### LINKED ISSUE #3866: [Feature] Expose retry behavior for template stanza
### Nomad version
Nomad v0.7.1

### Operating system and Environment details
Ubuntu 16.04.03 LTS

### Issue
When a running job encounters a template rendering failure due to Vault being inaccessible, the entire job is marked as "dead" and no rescheduling attempts are ever made. When this issue presents itself during a "net new" deploy, it's not a major deal as someone will be investigating the issue nearly immediately. However, when you have successfully deployed long-lived jobs that renew their secrets every N hours/days, this behavior becomes especially important. 

The desired behavior already exists in consul-template itself (for both consul communication and vault communication). See relevant config:
```
retry {
    # This enabled retries. Retries are enabled by default, so this is
    # redundant.
    enabled = true

    # This specifies the number of attempts to make before giving up. Each
    # attempt adds the exponential backoff sleep time. Setting this to
    # zero will implement an unlimited number of retries.
    attempts = 12

    # This is the base amount of time to sleep between retry attempts. Each
    # retry sleeps for an exponent of 2 longer than this base. For 5 retries,
    # the sleep times would be: 250ms, 500ms, 1s, 2s, then 4s.
    backoff = "250ms"

    # This is the maximum amount of time to sleep between retry attempts.
    # When max_backoff is set to zero, there is no upper limit to the
    # exponential sleep between retry attempts.
    # If max_backoff is set to 10s and backoff is set to 1s, sleep times
    # would be: 1s, 2s, 4s, 8s, 10s, 10s, ...
    max_backoff = "1m"
  }
```

Ideally, job operators should be able to configure these within the nomad template. I realize not everyone will want their jobs retrying indefinitely, but some jobs are mission-critical and requiring additional manual intervention *after* a Vault outage can be extremely costly. 

### Reproduction steps
1) Successfully deploy a job to Nomad that retrieves a short-lived secret from Vault via the "embedded consul-template" (example template below).
2) Make Vault inaccessible
3) Wait for attempt to renew secrets, resulting in a failure. Allocation transitions to "Failed" state.
4) Make Vault accessible again
5) Job remains dead and will require manual intervention.

### Example template
```
template {
        destination = "$${NOMAD_SECRETS_DIR}/server.bundle.pem"
        data = <<EOH
{{ $private_ip := env "NOMAD_IP_https" }}
{{ $ip_sans := printf "ip_sans=%s" $private_ip }}
{{ with secret "pki/us-west/issue/app" "common_name=app.service.consul" "alt_names=go-app.service.dc.consul" $ip_sans "format=pem" }}
{{ .Data.certificate }}
{{ .Data.issuing_ca }}
{{ .Data.private_key }}{{ end }}
EOH
      }
```

### LINKED ISSUE #11209: Nomad jobs should stay up when vault outage occurs
### Proposal
This issue aims to integration with vault. Running nomad jobs should stay to live as long as possible after vault outage.

### Use-cases
We have nomad cluster and vault in HA. Despite of clusters are scaled to cover minor outage of servers, sometimes it's unpredictable. After network issues on cloud provider side, our vault cluster was out of the service. 
Firstly nomad receive 503 response from vault and some short period restart all jobs with template stanza. All jobs ended in pending state until vault has been online. It not depends on change_mode in template stanza. 

### Attempted Solutions
Nomad could to switch to emergency mode when vault outage is detected. In this case common processes (ttl...) could be disabled and enabled again when vault will be available.

### LINKED ISSUE #11707: Expose `consul-template` configuration parameters for operators to tune.
### Use-cases

When Nomad clusters are experiencing Consul/Vault system degradation, or in situations where Operators have a higher tolerance for latency or connectivity loss (edge), a lack of control over `consul-template` fault tolerance behaviors can lead to a significant amount of allocation churn. `consul-template` exposes configuration options for fine-tuning retries, blocking queries, and startup fault tolerance. Having the ability to configure these options would be useful for Nomad Operators, especially when experiencing latency or instability when communicating with Consul.

### Proposal

Expose `consul-template` configuration parameters for operators to tune. The set of configuration options that will be exposed for `consul-template`, and their usage are:

- `max_stale` -  This is the maximum interval to allow "stale" data. By default, only the Consul leader will respond to queries. Requests to a follower or local agent will be forwarded to the leader. In large clusters with many requests, this is not as scalable, so this option allows any follower to respond to a query, so long as the last-replicated data is within these bounds. Higher values result in less cluster load but are more likely to have outdated data.
- `block_query_wait ` -  This is the amount of time to perform a blocking query. Many endpoints in Consul support a feature known as "blocking queries". A blocking query is used to wait for a potential change using long polling. This reduces the load on Consul by avoiding making new requests to Consul when nothing has changed.
- `wait` - This defines the minimum and maximum amount of time to wait for the cluster to reach a consistent state before rendering a template. This is useful to enable in systems that are experiencing a lot of flapping because it will reduce the number of times a template is rendered. This is configurable at _both the client and the task level_.  The task-level setting can be used to override the global setting.

```
  wait {
    enabled = true
    min     = "5s"
    max     = "90s"
  }
```

- `wait_bounds` - This is a Nomad-specific configuration that enables Nomad Operators to set client level constraints that set bounds on individual `jobspec` configuration. This setting defines lower and upper bounds for per-template `wait` configuration on a given client. If the individual template configuration has a `min` lower than `wait_bounds.min` or a `max` greater than the `wait_bounds.max`, the bounds will be enforced, and the template `wait` will be adjusted before being sent to `consul-template`.

 ```hcl
 wait_bounds {
   enabled = true
   min     = "5s"
   max     = "10s"
 }
   ```

- `consul_retry` - This controls the retry behavior when an error is returned from Consul. By default, Nomad will fail and reschedule an alloc when a template fails to render. This can lead to a significant `eval` -> `alloc` -> `template` render failure cycle in clusters where Consul is unstable. 

 ```hcl
  consul_retry {
    enabled = true
    # This specifies the number of attempts to make before giving up. Each
    # attempt adds the exponential backoff sleep time. Setting this to
    # zero will implement an unlimited number of retries.
    attempts = 12
    # This is the base amount of time to sleep between retry attempts. Each
    # retry sleeps for an exponent of 2 longer than this base. For 5 retries,
    # the sleep times would be: 250ms, 500ms, 1s, 2s, then 4s.
    backoff = "250ms"
    # This is the maximum amount of time to sleep between retry attempts.
    # When max_backoff is set to zero, there is no upper limit to the
    # exponential sleep between retry attempts.
    # If max_backoff is set to 10s and backoff is set to 1s, sleep times
    # would be: 1s, 2s, 4s, 8s, 10s, 10s, ...
    max_backoff = "1m"
  }

- `vault_entry` - This controls the retry behavior when an error is returned from Vault. By default, Nomad will fail and reschedule an alloc when a template fails to render. This can lead to a significant `eval` -> `alloc` -> `template` render failure cycle in clusters where Consul is unstable. 

 ```hcl
  vault_retry {
    # Same explanations as Consul
    enabled = true
    attempts = 12
    backoff = "250ms"
    max_backoff = "1m"
  }
```

## Related Issues

Closes #3866
Closes #2623

### PR DESCRIPTION
**Pull request by DerekStrickland**
See #11707 for a full explanation.

Expose Consul template configuration parameters for operators to tune at multiple levels:

- jobspec
  - wait - will override client default on a per template basis if set, or add if no client default set
- client
  - wait
  - wait_bounds - if set, will enforce upper and lower bounds on per template wait configured by jobspec authors
  - block_query_wait
  - max_stale - default to 5s
  - consul_retry
  - vault_retry

Closes #11707 
Closes #3866
Closes #2623
Closes #11209

### CODE REVIEW

**Review by tgross — COMMENTED** [review]
@DerekStrickland this is looking pretty good... I know these config PRs always end up being a huge slog because of all the code you have to touch! I've left a few questions about some of the notes you've left.

File: api/tasks.go
Comment (DerekStrickland) [reply api/tasks.go:778]: I had the same thought as I implemented, and you commenting on it, makes me feel like it should be implicit. I'll refactor to that.

File: client/allocrunner/taskrunner/template/template.go
Comment (DerekStrickland) [reply client/allocrunner/taskrunner/template/template.go:647]: For this review. I don't know if this is safe as is, or if my changes should result in us getting rid of this test-only setting, and just using the retry I've now defined.

File: client/config/config.go
Comment (DerekStrickland) [reply client/config/config.go:510]: Ah thanks. I think this comment is out of date. I'll update.

Comment (DerekStrickland) [inline client/config/config.go:510]:
```suggestion
// than Max the user specified Max.
```

File: command/agent/agent.go
Comment (DerekStrickland) [inline command/agent/agent.go:407]:
```suggestion
```

File: command/agent/job_endpoint.go
Comment (DerekStrickland) [reply command/agent/job_endpoint.go:1171]: Oops. This comment is out of date. Deleting.

Comment (DerekStrickland) [inline command/agent/job_endpoint.go:1171]:
```suggestion
```

File: nomad/structs/structs.go
Comment (DerekStrickland) [reply nomad/structs/structs.go:7567]: Probably not. Just kind of followed the pattern, but can definitely refactor to avoid the overhead.

File: website/content/docs/configuration/client.mdx
Comment (DerekStrickland) [reply website/content/docs/configuration/client.mdx:369]: Yes!  Super useful feedback. Thanks!

File: website/content/docs/configuration/client.mdx
Comment (DerekStrickland) [reply website/content/docs/configuration/client.mdx:363]: Thanks for the call out @tgross. 

Now that I am looking at the implementation and the original ticket a little deeper, I think I need to clarify the requirement. 

@schmichael  - Do we actually want to enforce a `5s` default for this or allow users to simply not set this value, and thus opt-in to the default `consul-template` behavior. The current logic only sets a value if the user specifies one, and this seems correct to me.

File: website/content/docs/configuration/client.mdx
Comment (DerekStrickland) [reply website/content/docs/configuration/client.mdx:411]: These are not default values, but rather example values. The Vault & Consul retry default configuration in `consul-template` are empty structs with no fields set. We don't have to provide that, because `consul-template` has a `Finalize` method that handles defaulting the values if callers don't set them.

Comment (DerekStrickland) [reply website/content/docs/configuration/client.mdx:411]: I'm going to resolve this conversation for now, but please re-open if I haven't addressed all the concerns here.

File: nomad/structs/structs.go
Comment (DerekStrickland) [reply nomad/structs/structs.go:7567]: Resolved in next push.

File: nomad/structs/diff_test.go
Comment (DerekStrickland) [reply nomad/structs/diff_test.go:7256]: Oops 😓   Fixed it in the next push.

File: client/config/config.go
Comment (DerekStrickland) [reply client/config/config.go:569]: I just felt a little apprehension passing pointer references around rather than copies of pointer values since it breaks encapsulation. I will go ahead and remove this comment since 

1. It behaves like the rest of the fields
2. It gets loaded only on agent startup
3. TaskTemplateManager only reads and never writes

File: command/agent/config_parse.go
Comment (DerekStrickland) [reply command/agent/config_parse.go:266]: yeah that makes sense.

File: website/content/docs/configuration/client.mdx
Comment (tgross) [reply website/content/docs/configuration/client.mdx:411]: LGTM

File: command/agent/config_parse.go
Comment (DerekStrickland) [reply command/agent/config_parse.go:266]: I just gave this a pass and ended up reverting. Moving this logic into the template runner resulted in having to make a call to a method that made a copy of this struct and then execute the `finalize` logic for every runner instance to avoid mutating shared state. This seemed wasteful. 

That said, I'd enjoy a discussion if you want/have time to have one. On the one hand, I don't love having domain logic related to a subsystem in the config loading/parsing code as that feels like a separate concern. On the other hand, I can understand an argument that that is actually the proper place as it localizes validation/initialization, and also because we only set these values to begin with because `hcl.decode` needs them to function properly. In the best case, we wouldn't even set them, and thus we wouldn't need to unset them.

File: .changelog/11606.txt
Comment (schmichael) [inline .changelog/11606.txt:2]:
```suggestion
template: Expose consul-template configuration options at the client level for `consul_retry`,
```

File: website/content/docs/configuration/client.mdx
Comment (schmichael) [reply website/content/docs/configuration/client.mdx]: Hm, I think a `5s` default is preferable because I suspect no one will think to configure this before they hit performance issues with Consul at which point they have to restart *all* of their Nomad agents to relieve some strain.

However, I hate changing default values like this in Z releases (where the 3 in 1.2.3 is the Z release), do you think we should hold this off until 1.3.0?

Comment (schmichael) [reply website/content/docs/configuration/client.mdx]: Nevermind! I misunderstood the *existing* default behavior: I assumed `max_stale` was *disabled* by default but it is [2s](https://github.com/hashicorp/consul-template/blob/d63e1925f4432ab5aadf452c02244392105b50ae/config/config.go#L539) (thanks to @DerekStrickland for the correction!).

Let's use consul-template's default then by default `max_stale` to an empty string.

File: api/tasks.go
Comment (DerekStrickland) [reply api/tasks.go]: Done

File: client/allocrunner/taskrunner/template/template.go
Comment (DerekStrickland) [reply client/allocrunner/taskrunner/template/template.go]: I refactored out the special testing only field and used the newly added RetryConfig since they were both used to manage the same Consul Template settings.

**Comment by DerekStrickland** [comment]
@tgross / @schmichael  - I believe this is ready for (hopefully!) final review.

File: client/allocrunner/taskrunner/template/template_test.go
Comment (tgross) [inline client/allocrunner/taskrunner/template/template_test.go:2088]: This works, but for future reference for tests like this where there's a lot of nested objects DRY-ing it up works against you a bit. It'd probably have been fine to have one big input `TaskTemplateManagerConfig` and one big "expected" `TemplateConfig` that you can compare with a single `require.Equal`.

File: website/content/docs/job-specification/template.mdx
Comment (tgross) [inline website/content/docs/job-specification/template.mdx:114]: This is a little confusing to me. With this wording I'd expect there to be a doc at `/docs/job-specification/task.mdx` as well, which isn't the case.

I'd leave out that "task level" sentence entirely and then reduce the wordiness around the client configuration option:

```suggestion
  because it will reduce the number of times a template is rendered. This setting 
  can be overridden by the [`client.template.wait_bounds`]. If the template
  configuration has a `min` lower than `client.template.wait_bounds.min` or a `max`
  greater than `client.template.wait_bounds.max`, the client's bounds will be enforced
  and the template `wait` will be adjusted before being sent to `consul-template`.
```

(and include the link reference to the client's `wait_bounds` setting at the bottom of the page)

File: website/content/docs/configuration/client.mdx
Comment (tgross) [inline website/content/docs/configuration/client.mdx:367]: We use the phrase "a lot of flapping" in this PR and although it's nice and pithy, it also feels like slang. Maybe replace with "where network connectivity to Consul is degraded"?

**Review by tgross — APPROVED** [review]
LGTM. I've left some final comments on documentation wording but once those are fixed I'm :+1: with merging this.

File: website/content/docs/configuration/client.mdx
Comment (DerekStrickland) [reply website/content/docs/configuration/client.mdx:367]: Done

**Pull request merged**
